README.md
# MySQL Database Schema For The Rowan Tree

Overview
--------

This project contains the MySQL Schema, and tools required to deploy and maintain the database for The Rowan Tree.

## Project Structure

```
./model
./stored_procs
./tables
./the dawn chant
```

* model - contains the MySQL Workbech model for the project.
* stored_procs - contains the create scripts for the stored procedures.
* tables - contains the create scripts for the tables.
* the dawn chant - contains the seed scripts required to bring the world into being.

Requirements
------------

* MySQL 5.7.1+ Server Instance (* or compatable)
* mysql client line tool (MySQL Workbench)
* mysql_config_editor
* PowerShell Core (6.0+)

Deploying
---------
The scripts to create the database require PowerShell 6.0+ (pwsh) and MySQL command line tool (mysql) to be present on the system.

Example invokcation of the database creation script

```
/$ pwsh 
PS> Import-Module ./Update-TRTDatabase.ps1
PS> $server = 'localhost'
PS> $port = '3306'
PS> $database = 'trtdb'
PS> $user = 'trt_service'
PS> Update-TRTDatabase -serverName $server -serverPort $port -databaseName $database -userName $user
```

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
MIT License

Copyright (c) 2018-2022 Joshua C. Burt

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
0.1.0
-----

- Added function to detect pre-existin login-path entry
- Added function to handle storing the credentials in the login-path
- Server port type changes from string to integer
- Added default username
- Added default login-path
- Added switch to drop and recreate the database
- Remove most of the showcmd output, it was generating noise in the result vaules of the functions (thanks powershell)
- Added more logic around detecting and setting the credentials
- Added sample script for consuming the cmdlet
- Removed definer from the stored procedure creation scripts
- Updated table statements, only create if missing
- Remove legacy code

0.0.2
-----

- Added some additional comments
- Added the ability to specify a port number for the remote server.
- Add missing delimiter changes within some of the stored procedures
- Cleaned up the stored procedures
- Added escaping to handle database names containing dashes '-'.

0.0.1
-----

- Initial Release

Update-Database.ps1

# If mysql is missing but the workbench is installed.
# 1) Then execute this to add the path to ~/.bash_profile:
#       echo 'export PATH=/usr/local/mysql/bin:$PATH' >> ~/.bash_profile
# 2) Then execute this, to reload the change:
#       . ~/.bash_profile

$server = 'localhost'
$username = 'root'
$password = ''
$database = 'dev_therowantree'


$mysql_cmd = "mysql --host $server --user $username --password=$password --database $database"

# Create the database
$database_create_statement = """CREATE DATABASE ``$database`` /*!40100 DEFAULT CHARACTER SET latin1 */;"""
$full_create = "mysql --host $server --user $username --password=$password --execute=$database_create_statement"
# Write-Output $full_create 
Invoke-Expression $full_create

Write-Output "Creating Tables.."
Get-ChildItem -Path ./tables | ForEach-Object -Process {
    $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
    # Write-Output $final_cmd
    Invoke-Expression $final_cmd
}

Write-Output('Creating Stored Procedures..')
Get-ChildItem -Path ./stored_procs | ForEach-Object -Process {
    $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
    # Write-Output $final_cmd
    Invoke-Expression $final_cmd
}

Write-Output('Adding seed data..')
Get-ChildItem -Path './the dawn chant' -Filter *.sql | ForEach-Object -Process {
    $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
    # Write-Output $final_cmd
    Invoke-Expression $final_cmd
}


Update-TRTDatabase.ps1
function Find-LoginPath {
    param (
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [String]$loginPath='local'
    )

    $foundLoginPath = $false
    $mysql_check_creds_cmd = "mysql_config_editor print --login-path $loginPath"

    try {
        $result = Invoke-Expression $mysql_check_creds_cmd
        if ($result.length -gt 0){
            $foundLoginPath = $true
        }
    }
    catch {
        Write-Error 'Find-LoginPath threw an exception!'
        Throw $_
    }
    return $foundLoginPath
}

function Set-MySQLCredentials {
    param (
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [String]$loginPath='local',

        [Parameter(Mandatory = $true)]
        [ValidateNotNull()]
        [String]$serverName,

        # Server Port For The MySQL Instance
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [Int]$serverPort=3306,

        # Server Port For The MySQL Instance
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [String]$userName='root'

        # The credentials object to connect to the database
        # [Parameter(Mandatory = $true)]
        # [PSCredential]$credentials
    )

    # $userName = $credentials.GetNetworkCredential().UserName
    # $userPassword = $credentials.GetNetworkCredential().Password

    # used to store the mysql credentials. the app doesn't allow passing the password through, thus the reason
    # we aren't using a pscredential object anymore.  Also the plain mysql cmd used below would through a warning
    # when passing the creds though.
    $mysql_remove_creds_cmd = "mysql_config_editor remove --login-path=$loginPath"
    $mysql_set_creds_cmd = "mysql_config_editor set --login-path=$loginPath --host=$serverName --port=$($serverPort.ToString()) --user=$userName --password"
    
    # Set our credentials
    try {
        Invoke-Expression "$mysql_remove_creds_cmd"
        Invoke-Expression "$mysql_set_creds_cmd"
        # $tmpfile = New-TemporaryFile
        # Set-Content -Value "$userPassword`r`n" -Path $tmpfile
        # Start-Process -FilePath "mysql_config_editor" -ArgumentList "set --login-path=$loginPath --host=$serverName --port=$serverPort --user=$userName --password" -RedirectStandardInput $tmpfile
        # Remove-Item $tmpfile
    }
    catch {
        Write-Error 'Set-MySQLCredentials threw an exception!'
        Throw $_
    }
}

function Update-TRTDatabase {
    # If mysql is missing but the workbench is installed.
    # 1) Then execute this to add the path to ~/.bash_profile:
    #       echo 'export PATH=/usr/local/mysql/bin:$PATH' >> ~/.bash_profile
    #    or  (macOS)
    #       echo 'export PATH=$PATH:/Applications/MySQLWorkbench.app/Contents/MacOS' >> ~/.bash_profile
    # 2) Then execute this, to reload the change:
    #       . ~/.bash_profile
    # If mysql_config_editor is missing but mysql server is installed.
    # 1) Then execute this to add the path to ~/.bash_profile:
    #       echo 'export PATH=/usr/local/mysql/bin:$PATH' >> ~/.bash_profile
    # 2) Then execute this, to reload the change:
    #       . ~/.bash_profile
    [CmdletBinding()]
    param (
        # Server Name For The MySQL Instance
        [Parameter(Mandatory = $true)]
        [String]$serverName,

        # Server Port For The MySQL Instance
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [Int]$serverPort=3306,

        # The database to create/update
        [Parameter(Mandatory = $true)]
        [ValidateNotNull()]
        [String]$databaseName,

        # The database to create/update
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [String]$userName='root',

        # The credentials object to connect to the database
        # [Parameter(Mandatory = $false)]
        # [PSCredential]$credentials,

        # The login path to store MySQL credentials at within .mylogin.cnf.
        [Parameter(Mandatory = $false)]
        [ValidateNotNull()]
        [String]$loginPath='local',

        [Switch]
        $displayCmd,

        # Controls updating entries within the .mylogin.cnf if they exist.
        [Switch]
        $updateCredentials
    )
        # Determine if we need to build the login-path within the local .mylogin.cnf.
        $loadCreds = Find-LoginPath $loginPath

        # if ($loadCreds -eq $false -and $credentials -eq $null){
        #     throw 'No credentials provided or stored!'
        # }

        if ($loadCreds -ne $true -or $updateCredentials -eq $true){
            $credObj = @{
                loginPath = $loginPath
                serverName = $serverName
                serverPort = $serverPort
                userName = $userName
            }
            Set-MySQLCredentials @credObj
        }

        # Build the begining of the sql cmds to execute.
        $databaseName = "$databaseName"
        $mysql_cmd = "mysql --login-path=$loginPath --database $databaseName"   

        # Create the database
        Write-Output 'Creating database..'
        $database_create_statement = """CREATE DATABASE ````$databaseName```` /*!40100 DEFAULT CHARACTER SET latin1 */;"""
        $full_create = "mysql --login-path=$loginPath --execute=$database_create_statement"
        if ($displayCmd) { Write-Output $full_create }
        Invoke-Expression $full_create

        Write-Output "Creating Tables.."
        Get-ChildItem -Path ./tables | ForEach-Object -Process {
            $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
            if ($displayCmd) { Write-Output $final_cmd }
            Invoke-Expression $final_cmd
        }

        Write-Output('Creating Stored Procedures..')
        Get-ChildItem -Path ./stored_procs | ForEach-Object -Process {
            $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
            if ($displayCmd) { Write-Output $final_cmd }
            Invoke-Expression $final_cmd
        }

        Write-Output('Adding seed data..')
        Get-ChildItem -Path './the dawn chant' -Filter *.sql | ForEach-Object -Process {
            $final_cmd = "$mysql_cmd -e ""source $($_.FullName)"" "
            if ($displayCmd) { Write-Output $final_cmd }
            Invoke-Expression $final_cmd
        }
}
